File: JavaScript/Core1/InventoryManagement/DailyDemand.js
class DailyDemand extends eVENT {
  constructor({ occTime, quantity, shop}) {
    super( occTime);
    this.quantity = quantity;
    this.shop = shop;
  }
  onEvent() {
    var q = this.quantity,
        prevStockLevel = this.shop.quantityInStock;
    // update lostSales if demand quantity greater than stock level
    if (q > prevStockLevel) {
      // increment the "nmrOfStockOuts" statistics variable by 1
      sim.stat.nmrOfStockOuts++;
      // increment the "lostSales" statistics variable by the missing quantity
      sim.stat.lostSales += q - prevStockLevel;
    }
    // update quantityInStock
    this.shop.quantityInStock = Math.max( prevStockLevel-q, 0);
    // schedule Delivery if stock level falls below reorder level
    if (prevStockLevel > this.shop.reorderLevel &&
        prevStockLevel - q <= this.shop.reorderLevel) {
      return [new Delivery({
        occTime: this.occTime + Delivery.sampleLeadTime(),
        quantity: this.shop.reorderUpToLevel - this.shop.quantityInStock,
        receiver: this.shop
      })];
    } else return [];  // no follow-up events
  }
  static sampleQuantity() {
    return math.getUniformRandomInteger( 5, 30);
  }
  static recurrence() {
    return 1;
  }
  createNextEvent() {
    return new DailyDemand({
      occTime: this.occTime + DailyDemand.recurrence(),
      quantity: DailyDemand.sampleQuantity(),
      shop: this.shop
    });
  }
}
DailyDemand.labels = {"quantity":"quant"};

File: Python/Core0/Inventory-Management/DailyDemand.py
import sys, os

module_path = os.path.abspath('OESpy-Core0/')
sys.path.insert(1, module_path)
from oes_foundation import eVENT

module_path = os.path.abspath('lib/')
sys.path.insert(1, module_path)
import util

from Delivery import Delivery

class DailyDemand( eVENT):
    def __init__(self, sim, quantity, shop, occTime=None, delay=None):
        super().__init__(sim, occTime, delay)
        self.quantity = quantity
        self.shop = shop

    def onEvent(self, sim):
        followupEvents = []
        q = self.quantity
        prevStockLevel = self.shop.quantityInStock
        # perform state changes
        self.shop.quantityInStock = max( prevStockLevel-q, 0)
        # update statistics
        if q > prevStockLevel:
            sim.stat["nmrOfStockOuts"] = sim.stat["nmrOfStockOuts"] + 1
            sim.stat["lostSales"] = sim.stat["lostSales"] + (q - prevStockLevel)
        # create follow-up events
        if prevStockLevel > self.shop.reorderLevel >= prevStockLevel - q :
            quantity = self.shop.targetInventory - self.shop.quantityInStock
            receiver = self.shop
            followupEvents.append( Delivery( sim, quantity, receiver, delay=Delivery.leadTime()))
        return followupEvents
    
    @staticmethod
    def quantity(): return util.getUniformRandomInteger(5, 30)
    
    @staticmethod
    def recurrence(): return 1
    
    def createNextEvent(self, sim):
        quantity = DailyDemand.quantity()
        shop = self.shop
        return DailyDemand( sim, quantity, shop, delay=DailyDemand.recurrence())
    
    def __str__(self):
        return "Demand@"+ str(self.occTime) + "{ quant:"+ str(self.quantity) +"}"
    